File: android/src/main/java/app/capgo/webviewversionchecker/WebviewVersionChecker.java
package app.capgo.webviewversionchecker;

import androidx.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;

final class WebviewVersionChecker {

    static final class CompatibilityEvaluation {

        final String state;
        final boolean isLatest;
        final String reason;

        CompatibilityEvaluation(String state, boolean isLatest, String reason) {
            this.state = state;
            this.isLatest = isLatest;
            this.reason = reason;
        }
    }

    @Nullable
    CompatibilityEvaluation evaluateCompatibility(
        @Nullable String currentVersionName,
        @Nullable Integer currentMajorVersion,
        @Nullable String latestVersion,
        @Nullable Integer minimumMajorVersion,
        @Nullable Double minimumDeviceSharePercent,
        @Nullable Double currentVersionSharePercent
    ) {
        if (minimumMajorVersion != null && currentMajorVersion != null && currentMajorVersion < minimumMajorVersion) {
            return new CompatibilityEvaluation(
                "outdated",
                false,
                "Installed WebView major version is below the configured minimum major version."
            );
        }

        if (minimumDeviceSharePercent != null && currentMajorVersion != null && currentVersionSharePercent != null) {
            boolean isLatest = currentVersionSharePercent >= minimumDeviceSharePercent;
            return new CompatibilityEvaluation(
                isLatest ? "latest" : "outdated",
                isLatest,
                isLatest
                    ? "Installed WebView major version satisfies the configured minimum device-share threshold."
                    : "Installed WebView major version is below the configured minimum device-share threshold."
            );
        }

        if (currentVersionName != null && !currentVersionName.isEmpty() && latestVersion != null && !latestVersion.isEmpty()) {
            int cmp = compareVersions(currentVersionName, latestVersion);
            boolean isLatest = cmp >= 0;
            return new CompatibilityEvaluation(
                isLatest ? "latest" : "outdated",
                isLatest,
                isLatest
                    ? "Installed WebView version is at or above the latest resolved version."
                    : "Installed WebView version is behind the latest resolved version."
            );
        }

        if (minimumMajorVersion != null && currentMajorVersion != null) {
            boolean isLatest = currentMajorVersion >= minimumMajorVersion;
            return new CompatibilityEvaluation(
                isLatest ? "latest" : "outdated",
                isLatest,
                isLatest
                    ? "Installed WebView major version satisfies the minimum configured major version."
                    : "Installed WebView major version is below the configured minimum major version."
            );
        }

        return null;
    }

    @Nullable
    Integer parseMajorVersion(@Nullable String version) {
        if (version == null || version.isEmpty()) {
            return null;
        }

        StringBuilder digits = new StringBuilder();
        for (int i = 0; i < version.length(); i++) {
            char c = version.charAt(i);
            if (Character.isDigit(c)) {
                digits.append(c);
            } else if (digits.length() > 0) {
                break;
            }
        }

        if (digits.length() == 0) {
            return null;
        }

        try {
            return Integer.parseInt(digits.toString());
        } catch (NumberFormatException error) {
            return null;
        }
    }

    int compareVersions(@Nullable String left, @Nullable String right) {
        List<Integer> leftParts = tokenize(left);
        List<Integer> rightParts = tokenize(right);

        int max = Math.max(leftParts.size(), rightParts.size());
        for (int index = 0; index < max; index++) {
            int a = index < leftParts.size() ? leftParts.get(index) : 0;
            int b = index < rightParts.size() ? rightParts.get(index) : 0;
            if (a > b) {
                return 1;
            }
            if (a < b) {
                return -1;
            }
        }

        return 0;
    }

    private List<Integer> tokenize(@Nullable String value) {
        List<Integer> out = new ArrayList<>();
        if (value == null || value.isEmpty()) {
            return out;
        }

        String[] rawParts = value.split("[^0-9]+");
        for (String part : rawParts) {
            if (part == null || part.isEmpty()) {
                continue;
            }
            try {
                out.add(Integer.parseInt(part));
            } catch (NumberFormatException ignored) {
                out.add(0);
            }
        }

        return out;
    }
}

File: ios/Sources/WebviewVersionCheckerPlugin/WebviewVersionChecker.swift
import Foundation
import UIKit

private let defaultIOSUpdateUrl = "https://support.apple.com/guide/iphone/update-ios-iph3e504502/ios"

@objc public class WebviewVersionChecker: NSObject {
    private let isoFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()

    func buildStatus(source: String, latestVersion: String?, minimumMajorVersion: Int?, updateUrl: String?) -> [String: Any] {
        let currentVersion = UIDevice.current.systemVersion
        let currentMajorVersion = parseMajorVersion(currentVersion)
        let normalizedLatestVersion = normalizeVersion(latestVersion)
        let latestMajorVersion = parseMajorVersion(normalizedLatestVersion)

        var status: [String: Any] = [
            "platform": "ios",
            "source": source,
            "checkedAt": isoFormatter.string(from: Date()),
            "currentVersion": currentVersion,
            "updateUrl": normalizedUpdateUrl(updateUrl)
        ]

        if let currentMajorVersion {
            status["currentMajorVersion"] = currentMajorVersion
        }

        if let normalizedLatestVersion {
            status["latestVersion"] = normalizedLatestVersion
        }

        if let latestMajorVersion {
            status["latestMajorVersion"] = latestMajorVersion
        }

        if let normalizedLatestVersion {
            let cmp = compareVersions(currentVersion, normalizedLatestVersion)
            let isLatest = cmp >= 0
            status["state"] = isLatest ? "latest" : "outdated"
            status["isLatest"] = isLatest
            status["reason"] = isLatest
                ? "Installed iOS WebKit version is at or above the requested latest version."
                : "Installed iOS WebKit version is behind the requested latest version."
            return status
        }

        if minimumMajorVersion != nil {
            status["state"] = "unknown"
            status["isLatest"] = false
            status["reason"] = "minimumMajorVersion applies to Android WebView only. Configure latestVersion on iOS to enforce a policy."
            return status
        }

        status["state"] = "unknown"
        status["isLatest"] = false
        status["reason"] = "iOS WebKit updates are tied to iOS updates. Configure latestVersion to enforce a policy."
        return status
    }

    func defaultPromptMessage(status: [String: Any]?) -> String {
        guard let status else {
            return "A newer iOS WebKit version may be available. Please update iOS for best compatibility."
        }

        let current = (status["currentVersion"] as? String) ?? "current version"
        let latest = (status["latestVersion"] as? String) ?? "the latest available version"

        return "Your iOS WebKit version (\(current)) may be outdated. Please update toward \(latest)."
    }

    func normalizedUpdateUrl(_ value: String?) -> String {
        let trimmed = normalizeVersion(value)
        return trimmed ?? defaultIOSUpdateUrl
    }

    func promptFingerprint(status: [String: Any]?, updateUrl: String) -> String {
        guard let status else {
            return updateUrl
        }

        let current = (status["currentVersion"] as? String) ?? ""
        let latest = (status["latestVersion"] as? String) ?? ""
        return "\(current)|\(latest)|\(updateUrl)"
    }

    private func parseMajorVersion(_ value: String?) -> Int? {
        guard let value, !value.isEmpty else {
            return nil
        }

        var digits = ""
        for character in value {
            if character.isNumber {
                digits.append(character)
            } else if !digits.isEmpty {
                break
            }
        }

        return Int(digits)
    }

    private func normalizeVersion(_ value: String?) -> String? {
        guard let value else {
            return nil
        }

        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    private func compareVersions(_ left: String, _ right: String) -> Int {
        let leftParts = tokenize(left)
        let rightParts = tokenize(right)
        let maxCount = max(leftParts.count, rightParts.count)

        for index in 0..<maxCount {
            let a = index < leftParts.count ? leftParts[index] : 0
            let b = index < rightParts.count ? rightParts[index] : 0
            if a > b {
                return 1
            }
            if a < b {
                return -1
            }
        }

        return 0
    }

    private func tokenize(_ value: String) -> [Int] {
        value
            .split(whereSeparator: { !$0.isNumber })
            .compactMap { Int($0) }
    }
}
